### ISSUE TITLE
support for custom field-mapping

### ISSUE THREAD

**Issue by pkalwar** [super-csv/super-csv#65]
Is it possible to indicate a custom field-mapping, in a map format or so?
Required if the csv headers are different then bean field-names.

This'd allow:
- less/no dependency on column position in csv
- csv can have any number of fields, in any order

Accordingly, cell-formatter array implementation would need to be changed.

Combined with Dozer, it'd be the most powerful tool.

Thanks for your great efforts.

**Comment by singularityfx** [comment]
Actually I was thinking about it too, and I believe it will be a great feature. I can work on this.

**Comment by kbilsted** [comment]
How will this be different than reading into a hash table and copy from
here into your object?

But sure let's see what you can come up with :-)
Den 18/08/2015 14.29 skrev "singularityfx" <notifications@github.com>:

**Comment by singularityfx** [comment]
@kbilsted 

I suggest to add additional overloads of CsvBeanReader.read() methods that accept name mapping and processors as maps instead of array:

`public <T> T read(final Class<T> clazz, final Map<String, String> nameMapping, final Map<String, CellProcessor> processors)`

`nameMapping`: key is actual column name in CSV file, value is bean field name
`processors`: key is the same, value is CellProcessor object

Compared to reading into a map and then instantiating a bean from it:

 - IMHO cleaner / more convenient / more declarative approach
 - You don’t have to provide array of all CSV column names, which is nice when CSV has thousands of columns. Also don’t have to provide CellProcessor array with thousands of nulls.
 - Method is independent on number / arrangement of columns in CSV file and names of columns unrelated to your bean. So same code can be used to instantiate beans even if there are changes in the CSV file.

Implementation:

1. The new `read()` methods just check arguments and forward the call to overloaded `readIntoBean()` methods

2. Overloaded `readIntoBean()` has 2 additional responsibilities:

 - Obtain actual CSV column names from the file using the existing `getHeader()` method (currently only used in examples and tests)

        if (csvHeader == null) {
   		 csvHeader = Arrays.asList(getHeader(true));
   	 }

 - Create processors array from map using CSV column names:

	private CellProcessor[] createProcessorsArrayFromMap(Map<String, CellProcessor> processorsMap) {
   	 CellProcessor[] processors = new CellProcessor[csvHeader.size()];
   	 
   	 for (int i = 0; i < csvHeader.size(); i++) {
   		 String columnName = csvHeader.get(i);
   		 if (processorsMap.containsKey(columnName)) {
   			 processors[i] = processorsMap.get(columnName);
   		 }
   	 }

   	 return processors;
    }

Other than that, the new readIntoBean() does basically the same job as existing one (calls `readRow()`, applies processors if any - in the end we have `processedColumns` populated with (processed) data from CSV row) and passes the map into new overload of `populateBean()`:

3. New `populateBean()`: here we cherry-pick the columns we need from `processedColumns` and put them into our bean.

     private <T> T populateBean(final T resultBean, final Map<String, String> nameMapping) {

   	 // map each column to its associated field on the bean
   	 // columns that are not associated with the bean are ignored
   	 for (String columnName : nameMapping.keySet()) {
   		 // find column number by looking up the columnName in the CSV header
   		 int i = csvHeader.indexOf(columnName);
   		 final Object fieldValue = processedColumns.get(i);

   		 // don't call a set-method in the bean if there is no result to store
   		 if( fieldValue == null ) {
   			 continue;
   		 }

   		 // invoke the setter on the bean
   		 Method setMethod = cache.getSetMethod(resultBean, nameMapping.get(columnName), fieldValue.getClass());
   		 invokeSetter(resultBean, setMethod, fieldValue);

   	 }

**Comment by ZioberMichal** [comment]
I think that it is a nice feature. Please, see my comment to implementation in PR.

**Comment by pkalwar** [comment]
Hello, I see this issue is still 'open', but a PR is in effect. 

Is the feature already available in latest release/trunk etc?

**Comment by ZioberMichal** [comment]
No, feature was not merged yet. We are still working on this feature to make it easy to use.

**Comment by kbilsted** [comment]
@singularityfx  @ZioberMichal  let's get going on this. How do you like my idea? It is simple. Work with eg beanreader and has a very terse syntax.

**Comment by kbilsted** [comment]
I think in order to go further with this idea, we must ensure that it works for all kinds of readers not only bean readers. Otherwise it won't properly fit it. If we can achieve that, I think we can slide it into the framework. :)

**Comment by pkalwar** [comment]
hello, any news yet?

**Comment by benlwong** [comment]
I currently have an exact need for this. I was evaluating a bunch of CSV libraries and would love to use Super CSV if it weren't for this feature. OpenCSV has this feature but it doesn't support something like CellProcessors. It would be awesome if someone resolve/finish implementing this issue.

**Comment by NanduKMG** [comment]
Hello guys, any update?
